src/scvi/model/base/_training_mixin.py

```diff
@@ -143,3 +158,195 @@ def train(
             **trainer_kwargs,
         )
         return runner()
+
+
+class SemisupervisedTrainingMixin:
+    _training_plan_cls = SemiSupervisedTrainingPlan
+
+    def _set_indices_and_labels(self):
+        """Set indices for labeled and unlabeled cells."""
+        labels_state_registry = self.adata_manager.get_state_registry(REGISTRY_KEYS.LABELS_KEY)
+        self.original_label_key = labels_state_registry.original_key
+        self.unlabeled_category_ = labels_state_registry.unlabeled_category
+
+        labels = get_anndata_attribute(
+            self.adata,
+            self.adata_manager.data_registry.labels.attr_name,
+            self.original_label_key,
+            mod_key=getattr(self.adata_manager.data_registry.labels, "mod_key", None),
+        ).ravel()
+        self._label_mapping = labels_state_registry.categorical_mapping
+
+        # set unlabeled and labeled indices
+        self._unlabeled_indices = np.argwhere(labels == self.unlabeled_category_).ravel()
+        self._labeled_indices = np.argwhere(labels != self.unlabeled_category_).ravel()
+        self._code_to_label = dict(enumerate(self._label_mapping))
+
+    def predict(
+        self,
+        adata: AnnOrMuData | None = None,
+        indices: Sequence[int] | None = None,
+        soft: bool = False,
+        batch_size: int | None = None,
+        use_posterior_mean: bool = True,
+    ) -> np.ndarray | pd.DataFrame:
+        """Return cell label predictions.
+
+        Parameters
+        ----------
+        adata
+            AnnData or MuData object that has been registered via corresponding setup
+            method in model class.
+        indices
+            Return probabilities for each class label.
+        soft
+            If True, returns per class probabilities
+        batch_size
+            Minibatch size for data loading into model. Defaults to `scvi.settings.batch_size`.
+        use_posterior_mean
+            If ``True``, uses the mean of the posterior distribution to predict celltype
+            labels. Otherwise, uses a sample from the posterior distribution - this
+            means that the predictions will be stochastic.
+        """
+        adata = self._validate_anndata(adata)
+
+        if indices is None:
+            indices = np.arange(adata.n_obs)
+
+        scdl = self._make_data_loader(
+            adata=adata,
+            indices=indices,
+            batch_size=batch_size,
+        )
+        y_pred = []
+        for _, tensors in enumerate(scdl):
+            inference_inputs = self.module._get_inference_input(tensors)  # (n_obs, n_vars)
+            data_inputs = {key: inference_inputs[key] for key in self.module.data_input_keys}
```

Comment:
@canergen This line is meant to allow for different numbers of measurements (e.g. mc + cov for BS-seq vs just RNA counts for RNA-seq) while reusing the majority of the `predict` function code.

It comes at the cost of requiring that a new field (`data_input_keys`) be specified in the module class, but this would allow for a lot more code re-use for semisupervised model/modules.